### DISCUSSION TITLE
Tailwind CSS with OpenMage

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#1825] [Ideas]
Until this year I used in Magento Bootstrap 3 and 4 Frameworks and I was very happy with the results. Mainly the responsive grid then the other components such as cards, buttons. In February I started evaluating Tailwind CSS and I can say that using it every day I got to appreciate it more and more. Tailwind is used in the class attribute of an html element. Also it gives you the ability to create more custom styles. It's absolutely fantastic. The Magento styles.css file is entirely removed.

It would be an interesting project if we could clean all the RWD and Base phtml files from inline styles and classes, Then add Tailwind to the html elements. At the end with Purgecss the final css file will be dramaticaly reduced to only the classes used in phtml files. What I really like is that I no longer have to create in the style file media sections for each screen size, also in the class of an element I can define the behavior for each resolution. Who is interested in playing with Tailwind can do so online here: https://play.tailwindcss.com/.

Bootstrap is still a popular framework, but not as flexible as Tailwind. Bootstrap themes tend to be similar. I don't dispute its advantages, the speed with which I can create visual elements in a page, but things are evolving and I think Tailwind is a winner. The debate is only about how you see the classes of an html element. Either you do it in a file as before, or inline. Then how easy can you make the changes. My personal opinion so far is that changes are much faster with Tailwind

**Comment by Flyingmana** [comment]
From a maintainer perspective I wonder, what could we do, to make maintaining such themes easier.
Getting rid of all the inline styles in not so much of an issue. But using css-framework specific class basically means, such a theme needs to overwrite nearly all of the templates.
For itself, this sounds not like a big issue, but we already had cases, where we needed to change/fix things in template files. And this usually would need to be applied to all themes, which did overwrite the same files.
It sounds wrong to need to overwrite the whole template for just adding class elements.
So I wonder, is there maybe an alternative we could establish for such themes?

Would it be sufficient, to allow enriching templates/blocks with class elements?
Do we need something more abstract, which allows to add classes or even more by html manipulation on block level?
(like add class "grid" to "div>div" of block "list.phtml")

**Comment by aterjung** [comment]
Hi,

I guess everybody has already noticed the hyvä M2 theme where tailwind is used. From my point of view, the way they did it is the only way to take profit out of it. Adding tailwind classes on to of the existing will add complexity, but it does not remove any. The m1 html is even More outdated than luma - I can’t imagine how that should bring any benefit.
Personally, i would like to see a new template total build from scratch with tailwind - and probably alpine JS just like hyvä for OpenMage. But that’s a hard pice of work to do, and nobody will use this in existing installations, cause it will break nearly every extension available.
In M2 there is some potential that extension devs will adapt there work if hyvä growth bigger, for OpenMage i am not to sure this could happen.

From a bigger view, if OpenMage derives to a more future proof code base, get rid of Zend1 for example, a theme like this without prototype and a small and modern html base could be very attractive for smaller customers. A lot of work.



Diplom Wirtschaftsinformatiker (FH)

**Comment by addison74** [comment]
Through the OpenMage project we try to take Magento 1 further, to be compatible with the new variants of PHP, MySQL, but we have never discussed about themes that are aging and require replacements. 

At first I rejected Tailwind CSS which seemed complicated and I didn't like loading the html elements with classes but the few tutorials on YouTube made me curious to dig deeper and realize that this option is worth taking account. The intention of the discussion was not to change the current functionality of OpenMage but to assess whether Tailwind CSS could be added to OpenMage as a separate theme for Frontend/Backend in the near future. 

There are some people who have asked for the implementation of Bootstrap for Backend, but as one who has used this framework and many others like Foundation intensely for a long time, I can say that it is not a very good idea. There are notable differences between Bootstrap and Tailwind CSS, which are very well explained if you search the Internet. The way Tailwind CSS was conceived as a first utility offers flexibility check its documentation here https://tailwindcss.com/docs/utility-first, 

Such a tool in the hands of an advanced UI/UX developer will lead to noticeable visual improvements in a new theme. I am at the evaluation stage for using it with OpenMage because a client asked me for a solution to replace the actual theme and there are very few updated themes on the market. Many of them have already remained anchored at the level of 2014 - 2017. 

Anyone with experience in creating new themes using modern technologies can express an opinion here.

**Comment by woutersamaey** [comment]
@ADDISON74 I am considering building this since quite some time, but would also like to do a few more breaking changes at the same time. So far, OpenMage has kept me back on the things I wanna do and I was thinking of forking OpenMage.
If OpenMage wants to be a drop-in replacement, that is fine, but now that Magento/Adobe let go of it, we can do amazing things and fix the shortcomings not even fixable in M2.

I have 3 pages worth of ideas for the future of OpenMage, but am still looking for funding/livelyhood.

**Reply by addison74** [reply]
Honestly, once Adobe made the announcement, many did not think that there was an alternative to Magento 1. Some were even forced to migrate spending a lot of money. I like Magento 2 but it still has a lot of issues and the team has a way of analyzing them not at all friendly.

We have this Discussion section where you can post all your ideas. In turn I released the Issue section of ideas and new features that I posted them in the Discussion section. I don't know how long this project will last but it's worth a try. 

The problem I see is the abandonment of some important developers to provide extensions and support for Magento 1. Otherwise, the code can be updated periodically according to PHP/MySQL changes. A skeleton for a theme can be made and from which new themes will result. Unfortunately we still have many issues and PR's open for a long time, some even important to be introduced in the code.

**Comment by aterjung** [comment]
Is there any reason to not start a new theme as a separate repo? An integration into OpenMage would be easy if it can be shown that this is possible and a great improvement, as log this will not break the legacy theme.

I would really like to be part of a development like this.

**Reply by woutersamaey** [reply]
This is what I meant in my last comment. Forking and doing some extras here and there perhaps. One example I'd like to add is a Menu interface, so changing / loading menu structures could be better organized.

**Reply by addison74** [reply]
I agree the first step to be a repo where we can find a skeleton based on RWD and Base from which all classes and inline styles have been removed. This way anyone can download it and start adding Tailwind classes. When someone makes a complete theme and wants to make it public he can give the news. The more people involved the better for the project.

I looked at your Hyva M2 suggestion and it's interesting what they achieved. Everything predefined in Bootstrap can be done with Tailwind CSS. There are even UI components for it that can be added, for example https://daisyui.com.

Obviously JavaScript libraries are also needed to get some features. For example zooming a product image. I have several proposals that can be implemented.

**Reply by aterjung** [reply]
I think RWD is not the right starting point, or do you recommend to copy RWD and wipe it out? Cleaning all inline styles and classes out of the template files should be possible with a short script. But it would not be very usefull. The HTML structure of all these themes is build in 2009 and RWD was quick and dirty approach make it a little bit more state of the art.
There are may Elements in alle M1 themes that you would not implement again in 2021. I would suggest to rebuild from scratch to gain all the profits of modern browsers.
In terms of JavaScript, i would also suggest to take a look at Alpine JS which is used in Hyvä. It is quite different, but promising.
Replacing Prototype would be a heavy lifting task. At least the the product option/configuration objects in the detail pages are quite complex. As far as i remember, these part are even mashed up with block classes and models...

**Reply by woutersamaey** [reply]
I agree with @aterjung . Starting fresh makes the most sense.

For the product options, I'd suggest a rebuild using Vue.js or something similar. It's far easier to maintain.

**Reply by addison74** [reply]
Ideally we would first update all the phtml and xml files copied from the RWD and Base directories, but this operation would mean starting from scratch with each file separately. I don't think we can achieve this without a solid team behind us and sharing tasks. For this reason I considered that the easiest way now would be to clean these files of all classes attached to html elements and inline styles. Once I get this skeleton I can play with Tailwind CSS. At a basic evaluation I think the hardest part will be until we get the structure of all pages then the rest of the blocks can be implemented much easier. We can go further by creating specific OpenMage classes that have the prefix "om-". These can be used without being repeated in various places.

**Reply by addison74** [reply]
@aterjung: the liberation from the Prototype library is a dream that unfortunately could not be realized in M1 being necessary a new version M2. I can't say that what they did in M2 is perfect but there are new elements closer to today's technologies. I'm not even a supporter of JQuery in the last year I was closer to vanilla JavaScript even if it was more complicated.

In any case we have to think about changing the frontend theme as long as OpenMage exists. I discovered old versions of X-Cart 4.1 and OpenCart with which I am still familiar and which had very attractive themes. It was a pleasure to spend some time there.

**Reply by aterjung** [reply]
@ADDISON74 if you mean as long as the goal is to be BC to the rest of the M1 Universe, then i would agree.

If cutting the roots is acceptable, then i think it can be done. It might hurt hard, but it's possible. Here is a startingpoint: https://blog.vianetz.com/2021/magento-remove-prototype/

If we are going to rebuild a themes HTML from scratch, thats not too strange ;-)

**Reply by addison74** [reply]
I can't agree with the removal of Prototype because there are many extensions that depend on this library. I recently installed one and it was extremely complex. Ideally we should remove it and be able to migrate these extensions but I am afraid that there is no more enthusiasm for such a thing. Look we are left with up to 20 people who are still interested in this project.

**Reply by aterjung** [reply]
yes, that is what i mean with backward compatibility. I would guess that 98% of all OM users use it as a replacement for M1 on legacy projects. There is no need for a OM version that breaks with everything. It is hard to imagine that anyone is willing to drop of all the extensions they are using.
This would lead to a new Shop System no longer compatible with Magento 1.

**Reply by woutersamaey** [reply]
A new shop system sounds good to me honestly. M2 is too heavy for most and M1 has a lot of stuff available. Sure, some adjustments and refreshing may be needed, but as a dev, I don’t mind.

**Reply by addison74** [reply]
If everyone has heard of React, Vue or Angular, it seems that some have seen the issues that each brings. There is more and more talk about Svelte these days. I didn't know about Alpine JS, but I quickly viewed some presentation on YouTube. Thanks for the tip. The creator who is a young full stack developer realized that something much easier was needed. 

A combination of Tailwind CSS and Alpine JS is interesting. It is fabulous not touching the css/javascript files and to do everything quickly directly in html. A lot of time was wasted moving around. We can make some drafts for clients extremely quickly without the need to use the traditional frameworks.

**Comment by Genaker** [comment]
There is implementation of the Magento open source Luma Them with the Tailwind styles added. No Magento junky Lees CSS and Hyva.
https://github.com/Genaker/Taillwind-Luna
Feel free to use it. there is also tool it shows diff of the core template and magento 2 enabled template. Fell free to use it is OSL 3.0.
This project is the **Tailwind** take on **Magento Luma**: same **Open Source** foundation and **child-of-Luma** inheritance for JS - just **Tailwind CSS** (plus small legacy  CSS like `checkout.css`) on top of Magento’s own templates and **Knockout / RequireJS** where the core still relies on them.

**Reply by addison74** [reply]
That theme cannot be used with OpenMage or Magento 1. It is for Magento 2 stores.

**Comment by Genaker** [comment]
Hi Addison,

It can be used; you just need to apply the same principles for M1. This is
more about the architecture than the code itself.

**Reply by sreichel** [reply]
Finally its moving the files to it correct place in OM and likly update some methods.

Thanks for sharing.
